proto/paginated_resource.go

```diff
@@ -0,0 +1,177 @@
+package proto
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+	"net/url"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/ably/ably-go/config"
+)
+
+// relLinkRegexp is the regexp that matches our pagination format
+var relLinkRegexp = regexp.MustCompile(`<(?P<url>[^>]+)>; rel="(?P<rel>[^"]+)"`)
+
+type errInvalidType struct {
+	typ reflect.Type
+}
+
+func (err errInvalidType) Error() string {
+	return "requested value of incompatible type: " + err.typ.String()
+}
+
+var (
+	msgType     = reflect.TypeOf((*[]*Message)(nil)).Elem()
+	statType    = reflect.TypeOf((*[]*Stat)(nil)).Elem()
+	presMsgType = reflect.TypeOf((*[]*PresenceMessage)(nil)).Elem()
+)
+
+// QueryFunc queries the given URL and gives non-nil HTTP response if no error
+// occurred.
+type QueryFunc func(url string) (*http.Response, error)
+
+// PaginatedResource represents a single page coming back from the REST API.
+// Any call to create a new page will generate a new instance.
+type PaginatedResource struct {
+	path       string
+	headers    map[string]string
+	links      []string
+	items      []interface{}
+	typedItems interface{}
+	typ        reflect.Type
+	query      QueryFunc
+}
+
+// NewPaginatedResource returns a new instance of PaginatedResource
+// It needs to be a struct implementing ResourceReader which in our case is rest.Client.
+func NewPaginatedResource(typ reflect.Type, path string, params *config.PaginateParams,
+	query QueryFunc) (*PaginatedResource, error) {
+	p := &PaginatedResource{
+		typ:   typ,
+		query: query,
+	}
+	builtPath, err := p.buildPaginatedPath(path, params)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := p.query(builtPath)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	p.path = builtPath
+	p.links = resp.Header["Link"]
+	v := reflect.New(p.typ)
+	if err = json.NewDecoder(resp.Body).Decode(v.Interface()); err != nil {
+		return nil, err
+	}
+	p.typedItems = v.Elem().Interface()
+	return p, nil
+}
+
+// Next returns the path to the next page as found in the response headers.
+// The response headers from the REST API contains a relative link to the next resource
+// (Link: <./path>; rel="next").
+func (p *PaginatedResource) Next() (*PaginatedResource, error) {
+	nextPath, ok := p.paginationHeaders()["next"]
+	if !ok {
+		return nil, errors.New("no next page after " + p.path)
+	}
+	nextPage, err := p.buildPath(p.path, nextPath)
+	if err != nil {
+		return nil, err
+	}
+	return NewPaginatedResource(p.typ, nextPage, nil, p.query)
+}
+
+// Items gives a slice of results of the current page.
+func (p *PaginatedResource) Items() []interface{} {
```

Comment:
We probably don't need to provide this function anymore since we have all types available.